rule_builder.py

```diff
@@ -0,0 +1,2069 @@
+    """A mapping of hash values to resolved rules"""
+
+    rule_item_dependencies: dict[str, set[int]]
+    """A mapping of item name to set of rule ids"""
+
+    rule_region_dependencies: dict[str, set[int]]
+    """A mapping of region name to set of rule ids"""
+
+    rule_location_dependencies: dict[str, set[int]]
+    """A mapping of location name to set of rule ids"""
+
+    rule_entrance_dependencies: dict[str, set[int]]
+    """A mapping of entrance name to set of rule ids"""
+
+    completion_rule: "Rule.Resolved | None" = None
+    """The resolved rule used for the completion condition of this world"""
+
+    true_rule: "Rule.Resolved"
+    """A pre-initialized rule for this world that always returns True"""
+
+    false_rule: "Rule.Resolved"
+    """A pre-initialized rule for this world that always returns False"""
+
+    item_mapping: ClassVar[dict[str, str]] = {}
+    """A mapping of actual item name to logical item name.
+    Useful when there are multiple versions of a collected item but the logic only uses one. For example:
+    item = Item("Currency x500"), rule = Has("Currency", count=1000), item_mapping = {"Currency x500": "Currency"}"""
+
+    rule_caching_enabled: ClassVar[bool] = True
+    """Enable or disable the rule result caching system"""
+
+    def __init__(self, multiworld: MultiWorld, player: int) -> None:
+        super().__init__(multiworld, player)
+        self.rules_by_hash = {}
+        self.rule_item_dependencies = defaultdict(set)
+        self.rule_region_dependencies = defaultdict(set)
+        self.rule_location_dependencies = defaultdict(set)
+        self.rule_entrance_dependencies = defaultdict(set)
+        self.true_rule = self.resolve_rule(True_())
+        self.false_rule = self.resolve_rule(False_())
+
+    @classmethod
+    def get_rule_cls(cls, name: str) -> type["Rule[Self]"]:
+        """Returns the world-registered or default rule with the given name"""
+        return CustomRuleRegister.get_rule_cls(cls.game, name)
+
+    @classmethod
+    def rule_from_dict(cls, data: Mapping[str, Any]) -> "Rule[Self]":
+        """Create a rule instance from a serialized dict representation"""
+        name = data.get("rule", "")
+        rule_class = cls.get_rule_cls(name)
+        return rule_class.from_dict(data, cls)
+
+    def resolve_rule(self, rule: "Rule[Self]") -> "Rule.Resolved":
+        """Returns a resolved rule registered with the caching system for this world"""
+        resolved_rule = rule.resolve(self)
+        resolved_rule = self.simplify_rule(resolved_rule)
+        return self.get_cached_rule(resolved_rule)
+
+    def get_cached_rule(self, resolved_rule: "Rule.Resolved") -> "Rule.Resolved":
+        """Returns a cached instance of a resolved rule based on the hash"""
+        rule_hash = hash(resolved_rule)
+        if rule_hash in self.rules_by_hash:
+            return self.rules_by_hash[rule_hash]
+        self.rules_by_hash[rule_hash] = resolved_rule
+        return resolved_rule
```

Comment:
So this is for memory? I don't like that this isn't doing an `__eq__` check. I guess couldn't this just be a basic dict called `rules_cache` and then do `self.rules_cache[resolved_rule] = resolved_rule`?

Also, couldn't this be a global cache by moving it to the Resolved metaclass?